### PR TITLE
Don't load providers at startup

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This changes the provider registry to no longer load all the providers
from the old state on startup (in `NewRegistry`) instead the load logic
has been moved to the `Same` method. The step_executor and
step_generator have been fixed up to ensure that for cases where a
resource might not have had it's provider created yet (i.e. for DBR'ing
the old version of a resource, for refreshes or deletes) they ask the
`Deployment` to look up the provider in the old state and `Same` it in the
registry.

All of the above means we only load providers we're going to use (even
taking --targets into account).

One fix mot done in this change is to auto-update providers for deletes.
That is given a program state with two resources both using V1 of a
provider, if you run the program to update one of those resource to use
V2 of the provider but to delete the other resource currently we'll
still load V1 to do that delete. It _might_ be possible (although this
is definitly questionable) to see that another resource changed it's
provider from V1 to V2 and to just assume the same change should have
happened to the deleted resource.
This could be helpful for not loading old provider versions at all, but
can be done in two passes now pretty easily. Just run `up` without any
program changes except for the SDK version bump to update all the
provider references to V2 of the provider, then do another `up` that
deletes the second resource.
    
Fixes https://github.com/pulumi/pulumi/issues/12177.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/resource/deploy/providers/registry.go
Comment (abhinav) [inline pkg/resource/deploy/providers/registry.go:469]: We only have a read lock here. We cannot mutate. Need to switch to a write lock.
As optimization for the common case, we should keep the read lock up until the check-map-and-early-return.
After that, unlock, get a write lock, check again (in case another goroutine was also trying to create that provider), and if not, do the rest of this.

File: pkg/resource/deploy/providers/registry.go
Comment (abhinav) [inline pkg/resource/deploy/providers/registry.go:418]: Same feels like a strange place to do this. I'm not quite familiar with this code, but would GetProvider be a more appropriate place to do this?
Yeah, GetProvider only takes a reference, though, so it won't have access to state information.
One possibility is still passing `prev` into NewRegistry, and recording old provider states for later.
GetProvider can then do: "if have provider, return. otherwise, if have old state, build from that."
A nice side effect of doing that will be that if any code today assumes NewRegistry followed by GetProvider works, it'll continue to work with the same information.

**Review by abhinav — COMMENTED** [review]
I'm unsure about doing this in Same step, but I'm also not very familiar with that part of the workflow. Just based on how data flows here, I have suggested an alternative. If Same step is the correct place for this, I'd like to get more details on Monday.

---

Unrelated to ^: I have pushed a small fixup commit. Please pull to your local copy before next push.

**Comment by pgavlin** [comment]
> That is given a program state with two resources both using V1 of a
provider, if you run the program to update one of those resource to use
V2 of the provider but to delete the other resource currently we'll
still load V1 to do that delete.

Just for context: this behavior was intentional. Historically we’ve seen some issues with deleting old resources using newer providers if there have been breaking changes between versions. I suspect that it is not safe to delete an older version of a resource with a newer provider in the general case. If we had an explicit state migration step we might be able to make that work, though.

**Comment by Frassle** [comment]
> Just for context: this behavior was intentional.

Would we want to keep that behavior for replacements as well then? I.e. if you do an Update it just uses the new provider, but if you do a replace (DBR or RBD) the old resource gets Deleted with the old provider.
I had put in a change as part of this so that replacement deletes would use the new provider, but similar to the delete point you were replying to if you had issues using the old provider you could just do a no-op update to get the new provider then do the replaces in a second update. So doesn't seem to painful to keep loading old providers by default here.

File: pkg/resource/deploy/step_generator.go
Comment (Frassle) [reply pkg/resource/deploy/step_generator.go:1023]: No we're not sure, see @pgavlin [comment](https://github.com/pulumi/pulumi/pull/12657#issuecomment-1518949137). We might want to change this back to keep using the old provider before merging.

File: pkg/resource/deploy/source_query.go
Comment (Frassle) [reply pkg/resource/deploy/source_query.go:54]: I don't think anyone really uses query mode, I should do a sanity pass that this doesn't completely break it though.

File: pkg/engine/lifecycletest/provider_test.go
Comment (Frassle) [reply pkg/engine/lifecycletest/provider_test.go:766]: Ah oops no this should be set lower down

**Comment by pgavlin** [comment]
> > Just for context: this behavior was intentional.
> 
> Would we want to keep that behavior for replacements as well then? I.e. if you do an Update it just uses the new provider, but if you do a replace (DBR or RBD) the old resource gets Deleted with the old provider. I had put in a change as part of this so that replacement deletes would use the new provider, but similar to the delete point you were replying to if you had issues using the old provider you could just do a no-op update to get the new provider then do the replaces in a second update. So doesn't seem to painful to keep loading old providers by default here.

Yep, I would imagine that the old provider would be used for the deletion in a DBR/RDB scenario as well.

We probably have an issue open somewhere for explicit state migration that we should address.

File: pkg/resource/deploy/providers/registry.go
Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go:418]: Could do, would shrink this down to just dealing with aliases again. I'd lent away from that style because doing it this way instead required us to more strictly think about when we were creating providers (You _had_ to run a Create/Update/Same for the provider to start up). Things get a lot shorter code wise if GetProvider just does it, but is maybe less clear what's causing providers to be created.

So 🤷 either works, I'm happy to rejig if others feel doing this in GetProvider would feel better.

File: pkg/resource/deploy/providers/registry.go
Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go:469]: Ah good catch

File: pkg/resource/deploy/providers/registry.go
Comment (abhinav) [reply pkg/resource/deploy/providers/registry.go:418]: GetProvider felt like the right place because Same looks like it should be a read-only operation. I can see value in the explicitness requirement. I'll defer to your choice between the two.

Thanks for the keywords. They helped dig around and understand this better. Looks like Check is the other place that builds the provider (which, I'm guessing is called before Create or Update because of the assertion in them—I think the state machine here is not super clear, but that's an adjacent problem). I looked to see if there was a way to share any code between Check and Same for provider creation. Some of it could be shared, but the validation that happens before adding to the providers map differs (CheckConfig versus Configure).

Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go:418]: > Same looks like it should be a read-only operation.

So should Check really if your thinking of these like cloud resources, but they aren't so I'm not sure this really needs to hold. I'm leaning towards the explictness but we could pretty easily rework this in the future, it's not a hard thing to unpick.

File: pkg/resource/deploy/providers/registry.go
Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go]: Interestingly this was previously unlocked but I guess hit rarely enough that we never trigger a concurrent write bug.

File: pkg/resource/deploy/providers/registry.go
Comment (Frassle) [reply pkg/resource/deploy/providers/registry.go:418]: Actually I think because of the way the other methods (Check/Diff/Update/Create) use GetProvider this does need to be explicit.

File: pkg/resource/deploy/step_generator.go
Comment (Frassle) [reply pkg/resource/deploy/step_generator.go]: Changed this back

**Comment by Frassle** [comment]
> We probably have an issue open somewhere for explicit state migration that we should address.

Not sure that's strictly needed after this, it'll just be a case of doing a no changes `up`.

**Review by justinvp — APPROVED** [review]
The changes in the PR LGTM.

One thing I don't understand: in order to fix #12177, don't we need to change the implementation and/or callers of `installPlugins`:

https://github.com/pulumi/pulumi/blob/3e69a0bbc30380646001f7450c780a33189c694b/pkg/engine/update.go#L215-L263

?

File: pkg/resource/deploy/deployment.go
Comment (abhinav) [inline pkg/resource/deploy/deployment.go:503]:
```suggestion
		// We need to create the provider in the registry, find its old state and just "Same" it.
```

**Review by abhinav — APPROVED** [review]

**Comment by Frassle** [comment]
> One thing I don't understand: in order to fix https://github.com/pulumi/pulumi/issues/12177, don't we need to change the implementation and/or callers of installPlugins:

We _could_, but I think that's actually fine to keep. In the happy case it will pre-fetch the plugins needed for the run. In the error case we don't hard error if any of that set fails to install. So the common error cases of like "old version of plugin doesn't have an architecture to match the current machine" doesn't really matter as long as we then run an update such that it doesn't actually need that.

**Comment by Frassle** [comment]
bors merge

**Pull request merged**
